eng/tools/azperf/README.md

```diff
@@ -0,0 +1,120 @@
+# Performance Testing Framework
+The `azperf` CLI tool provides a singular framework for writing and running performance tests.
+
+## Default Command Options
+
+| Flag | Short Flag | Default Value | Variable Name | Description |
+| -----| ---------- | ------------- | ------------- | ----------- |
+| `--duration` | `-d` | 10 seconds | internal.Duration (`int`) | How long to run an individual performance test |
+| `--iterations` | `-i` | 3 | internal.Iterations (`int`) | How many iterations of a performance test to run |
+| `--testproxy` | `-x` | N/A | internal.TestProxy (`string`) | Whether to run a test against a test proxy. If you want to run against `https` specify with `--testproxy https`, likewise for `http`. If you want to run normally omit this flag |
+| `--warmup` | `-w` | 3 seconds| internal.WarmUp (`int`) | How long to allow the connection to warm up. |
+| `--timeout` | `-t` | 10 seconds| internal.TimeoutSeconds (`int`) | How long to allow an operation to block. |
+
```

Comment:
I'm not very familiar with how they work in go, but "parallelism" and "sync vs async" are important concepts in the other languages perf frameworks.

For "sync vs async", other languages like .NET, Java, and Python expose both sync and async APIs in our SDK clients, and `Run()` vs `RunAsync()` methods on the `PerfTest` base class.  This allows us to test both flavor of APIs, compare their performance, etc.  JS is async-only.

The "parallel" parameter controls how many instances of the test class will be executed in parallel.  For example, if "parallel=4", the perf framework will:

1. Create 4 instances of the test class
2. Call `GlobalSetup()` on the first instance
3. Call `Setup()` on each instance (in parallel)
4. Call `Run/RunAsync()` on each instance in a loop (in parallel)
5. Call `Cleanup()` on each instance (in parallel)
6. Call `GlobalCleanup()` on the first instance

Reply:
Go is a bit different than the other languages. There are no explicit async APIs in our clients. Async programming is done via [goroutines](https://gobyexample.com/goroutines), but this is done entirely on the client side. I can add parallel testing for this by creating the `goroutines` manually but this will just test the sync APIs. I think for the first iteration of the performance framework, doing this without `goroutines` is ideal and I can add them later if they are necessary.